### PR TITLE
CLOUDP-82251: Refactor scram logic to be importable from enterprise

### PR DESCRIPTION
**Pull request by chatton**
### All Submissions:

* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

This PR refactors the code that configures SCRAM authentication in such a way that makes it importable in the enterprise code base. This is done by defining an interface and a set of types which will be built in both projects. The code that will be shared will configure the automation config and also generate / ensure that the Secrets which store the user credentials are created.

### CODE REVIEW

File: pkg/authentication/scram/mock_types_test.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
+package scram
```
Comment (chatton) [inline pkg/authentication/scram/mock_types_test.go:1]: this defines a set of mock implementations to break any dependency on the `mdbv1` package.

File: pkg/authentication/scram/scram.go [hunk 2]
```diff
@@ -28,37 +27,108 @@ const (
...
+	PasswordSecretName string
+
+	// ScramCredentialsSecretName returns the name of the secret which stores the generated credentials
+	// for this user. These credentials will be generated if they do not exist, or used if they do.
+	// Note: there will be one secret with credentials per user created.
+	ScramCredentialsSecretName string
+}
+
+// Options contains a set of values that can be used for more fine grained configuration of authentication.
+type Options struct {
```
Comment (chatton) [inline pkg/authentication/scram/scram.go:79]: Options contains all the fields that different between enterprise/community

File: pkg/authentication/scram/scram.go [hunk 3]
```diff
@@ -28,37 +27,108 @@ const (
...
+	// Name is the name of the role.
+	Name string
+
+	// Database is the database this role applies to.
+	Database string
+}
+
+// User is an interface which must be implemented by a type which will map to an
+// automationconfig.MongoDBUser.
+type User struct {
```
Comment (chatton) [inline pkg/authentication/scram/scram.go:56]: contains information for the User that will be in the automation config and values related to the password secret

File: pkg/authentication/scram/scram.go [hunk 4]
```diff
@@ -28,37 +27,108 @@ const (
 	sha256StoredKeyKey = "sha-256-stored-key"
 )
 
-// EnsureScram will configure all of the required Kubernetes resources for SCRAM-SHA to be enabled.
+// Configurable is an interface which any resource which can configure ScramSha authentication should implement.
+type Configurable interface {
```
Comment (chatton) [inline pkg/authentication/scram/scram.go:31]: this will be implemented by MongoDBCommunity + AppDB

File: controllers/replica_set_controller.go [hunk 5]
```diff
@@ -540,10 +536,15 @@ func (r ReplicaSetReconciler) buildAutomationConfigSecret(mdb mdbv1.MongoDBCommu
 		return corev1.Secret{}, errors.Errorf("could not read existing automation config: %s", err)
 	}
 
+	auth := automationconfig.Auth{}
+	if err := scram.Enable(&auth, r.client, mdb); err != nil {
```
Comment (chatton) [inline controllers/replica_set_controller.go:540]: this enables scram + creates all k8s resources required

File: pkg/authentication/scram/scram.go [hunk 6]
```diff
@@ -28,37 +27,108 @@ const (
...
+
+	// AutoAuthMechanism is the desired authentication mechanism that the agents will use.
+	AutoAuthMechanism string
+}
+
+// Enable will configure all of the required Kubernetes resources for SCRAM-SHA to be enabled.
 // The agent password and keyfile contents will be configured and stored in a secret.
 // the user credentials will be generated if not present, or existing credentials will be read.
-func EnsureScram(secretGetUpdateCreateDeleter secret.GetUpdateCreateDeleter, secretNsName types.NamespacedName, mdb mdbv1.MongoDBCommunity) (automationconfig.Modification, error) {
+func Enable(auth *automationconfig.Auth, secretGetUpdateCreateDeleter secret.GetUpdateCreateDeleter, mdb Configurable) error {
```
Comment (chatton) [inline pkg/authentication/scram/scram.go:99]: we mutate this instead of returning function

File: pkg/authentication/scram/scram_enabler.go [hunk 7]
```diff
@@ -47,9 +61,34 @@ func enableAgentAuthentication(auth *automationconfig.Auth, agentPassword, agent
...
+	for _, authMode := range opts.AutoAuthMechanisms {
+		if !contains.String(auth.DeploymentAuthMechanisms, authMode) {
+			auth.DeploymentAuthMechanisms = append(auth.DeploymentAuthMechanisms, authMode)
+		}
+	}
+}
+
+// validateScramOptions validates that all the required fields have
+// a non empty value.
+func validateScramOptions(opts Options) error {
```
Comment (chatton) [inline pkg/authentication/scram/scram_enabler.go:74]: before we specified valid options directly, now that they are configurable they need to be validated

File: pkg/authentication/scram/scram_test.go [hunk 8]
```diff
@@ -39,32 +40,6 @@ const (
...
-}
-
-func (c mockSecretGetUpdateCreateDeleter) GetSecret(objectKey client.ObjectKey) (corev1.Secret, error) {
-	if s, ok := c.secrets[objectKey]; !ok {
-		return corev1.Secret{}, notFoundError()
-	} else {
-		return s, nil
-	}
-}
-
```
Comment (chatton) [inline pkg/authentication/scram/scram_test.go:67]: this was moved to a different file

File: api/v1/mongodbcommunity_types.go [hunk 9]
```diff
@@ -364,6 +358,42 @@ type MongoDBCommunity struct {
...
+		AutoAuthMechanisms: []string{scram.Sha256},
+		AgentName:          scram.AgentName,
+		AutoAuthMechanism:  scram.Sha256,
+	}
+}
+
+// GetScramUsers converts all of the users from the spec into users
+// that can be used to configure scram authentication.
+func (m MongoDBCommunity) GetScramUsers() []scram.User {
+	var users []scram.User
```
Comment (bznein) [inline api/v1/mongodbcommunity_types.go:376]: [nit] can we pre-allocate this and avoid `append`?

Comment (bznein) [reply api/v1/mongodbcommunity_types.go:376]: (the same for roles below)

File: pkg/authentication/scram/scram.go [hunk 10]
```diff
@@ -28,37 +27,108 @@ const (
...
+type Role struct {
+	// Name is the name of the role.
+	Name string
+
+	// Database is the database this role applies to.
+	Database string
+}
+
+// User is an interface which must be implemented by a type which will map to an
+// automationconfig.MongoDBUser.
```
Comment (bznein) [inline pkg/authentication/scram/scram.go:55]: This comment seems wrong? This is not an interface

File: pkg/authentication/scram/scram_enabler.go [hunk 11]
```diff
@@ -47,9 +61,34 @@ func enableAgentAuthentication(auth *automationconfig.Auth, agentPassword, agent
...
+	}
+	if opts.KeyFile == "" {
+		errs = multierror.Append(errs, errors.New("KeyFile must be specified"))
+	}
+
+	if errs != nil {
+		return errs
+	}
+
+	return nil
```
Comment (bznein) [inline pkg/authentication/scram/scram_enabler.go:93]: [nit] if possible, just `return errs`?

**Review by bznein — COMMENTED** [review]
Looks great, comments about real minor things

File: api/v1/mongodbcommunity_types.go (hunk 9 above)
Comment (chatton) [reply api/v1/mongodbcommunity_types.go:376]: good idea, will update!

File: pkg/authentication/scram/scram.go (hunk 10 above)
Comment (chatton) [reply pkg/authentication/scram/scram.go:55]: thank you stale comment!

**Comment by chatton** [comment]
@bznein addressed your feedback! Please take another look when you have some time.

**Review by bznein — APPROVED** [review]
LGTM! 🥳

**Pull request merged**